fontforge/splinefit.c

```diff
@@ -733,7 +805,223 @@ return( SplineMake3(from,to));
+    BasePoint bunit = (BasePoint) { BPDot(BPRev(ftunit), tounit),BPCross(ftunit, tounit) }; /* normed direction at "to" */
+    if ( aunit.y < 0 ) { /* normalize aunit.y to >= 0: */
+		aunit.y = -aunit.y;
+		bunit.y = -bunit.y;
+		m = -m;
+		f = -f;
+	}
+	/* calculate the Tunni point (where the tangents at "from" and "to" interesect) */
+	bigreal aMax = 100; /* maximal value that the handle a can reach up to the Tunni point, 100 is really long */
+	bigreal bMax = 100; /* maximal value that the handle b can reach up to the Tunni point, 100 is really long */
+	sab = aunit.y*bunit.x+aunit.x*bunit.y; 
+	if (sab != 0) { /* if handles not parallel */
+		aMax = bunit.y/sab;
+		bMax = aunit.y/sab;
+		if ( aMax < 0 ) {
+				aMax = 100;
+		}
+		if ( bMax < 0 ) {
+				bMax = 100;
+		}
+	}
+	/* start approximation by solving the quartic equation */
+	sasa = aunit.y*aunit.y; /* reducing the multiplications */
+	Quartic quad;
+	if ( (aunit.x == -bunit.x && aunit.y == bunit.y) || (aunit.x == bunit.x && aunit.y == -bunit.y) ) { /* handles are parallel */
+		quad.a = 0;
+		quad.b = 0;
+		quad.c = -9*aunit.x*sasa;
+		quad.d = 6*aunit.y*(4*aunit.y+5*aunit.x*f);	
+		quad.e = 10*((42*m-25*f)*aunit.y-25*aunit.x*f*f);
+	} else { /* generic situation */
+		quad.a = -9*aunit.x*(((2*bunit.y*bunit.x*aunit.x+aunit.y
+				*(2*bunit.x*bunit.x-1))*aunit.x-2*bunit.y*bunit.x)
+				*aunit.x-bunit.x*bunit.x*aunit.y);
+		quad.b = 12*((((bunit.x*(30*f*bunit.x-bunit.y)-15*f)
+				*aunit.x+2*aunit.y-bunit.x*aunit.y*(bunit.x+30*f*bunit.y))
+				*aunit.x+bunit.x*(bunit.y-15*f*bunit.x))
+				*aunit.x-aunit.y*bunit.x*bunit.x);	
+		quad.c = 12*((((70*m+15*f)*bunit.y*bunit.y+bunit.x
+				*(9*bunit.y-70*bunit.x*m-5*bunit.x*f))
+				*aunit.x-5*aunit.y*bunit.y*(3*bunit.y-4*bunit.x
+				*(7*m+f)))*aunit.x-bunit.x*(9*bunit.y-70*bunit.x*m-5*bunit.x*f));
+		quad.d = 16*(((12*aunit.y-5*aunit.x*(42*m-17*f))*bunit.y
+				-70*bunit.x*(3*m-f)*aunit.y-75*aunit.x*bunit.x*f*f)
+				*bunit.y-75*bunit.x*bunit.x*f*f*aunit.y);
+		quad.e = 80*bunit.y*(42*bunit.y*m-25*f*(bunit.y-bunit.x*f));
+	}
+	extended solutions[4] = {-999999,-999999,-999999,-999999};
+	_QuarticSolve(&quad,solutions);
+	extended abSolutions[10][2]; /* there are at most 4+4+1+1=10 solutions of pairs of a and b (quartic=0,derivative=0,b=0.01,a=0.01) */
+	numberOfSolutions = 0;
+	extended a,b;
+	for( int i = 0; i < 4; i++ ){
+		a = solutions[i];
+		if ( a >= 0 && a < aMax ) {
+			b = (20*f-6*a*aunit.y)/(3*(2*bunit.y-a*sab));
+			if ( b >= 0 && b < bMax ) {
+				abSolutions[numberOfSolutions][0] = a;
+				abSolutions[numberOfSolutions++][1] = b;
+			}
+		}
+	}
+	/* and now again for the derivative (of m not of the upper quartic): */
+	if ( (aunit.x == -bunit.x && aunit.y == bunit.y) || (aunit.x == bunit.x && aunit.y == -bunit.y) ) { /* handles are parallel */
+		quad.a = 0;
+		quad.b = 0;
+		quad.c = 0;
+		quad.d = -3*aunit.x*aunit.y;
```

Comment:
I can't seem to match this to the comment:

```c
/* 0 = -9*ca*sa^3 * a^3  */
/* + 6*sa^2*(4*sa+5*ca*f) * a^2 */
/* + 10*sa*((42*m-25*f)*sa-25*ca*f^2). */
```

If this were matching, wouldn't it be

```c
		quad.c = -9*aunit.x*sasa*aunit.y;
		quad.d = 6*sasa*(4*aunit.y+5*aunit.x*f);	
		quad.e = 10*((42*m-25*f)*aunit.y-25*aunit.x*f*f);
```

?